feat: Add navigation layer

thehale opened the pull request:
Adds a modular structure to the WearOS UI, including navigation, screens, layouts, components, and remember functions.

With these improvements, developers should have an easier time working on independent parts of the application without creating tons of merge conflicts.

thehale committed:
feat: Improve debug logging in `Messenger`

Now all connected devices are listed, to help narrow down
why messages may not get through to the phone.

thehale committed:
fix: Remove `name` from `CurrentActivity` in tests

The `name` parameter no longer exists on the DTO, so it was
breaking the tests.

thehale committed:
fix: Correct `map` test in `UnCoveredRangesMapper`

The test was failing erroneously due to an incorrectly written
Range in the expected output.

By making the expected output follow the same format as the
other tests, the failing test began passing once again.

thehale committed:
feat(wear): Add navigation layer

This is the last big commit for creating the initial, guiding
project structure.

Replaces the monolithic `MainActivity` with a variety of
narrowly focused Composables which assemble nicely into
a full application.

In particular, this commit shows off the ability to press an activity
which opens a new screen with chips for that activity's tags.

A major milestone towards #5

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/thehale/SimpleTimeTracker-WearOS?pullRequest=10) <br/>All committers have signed the CLA.

kantahrek commented on wear/src/main/java/com/example/util/simpletimetracker/presentation/remember/activities.kt:
```diff
@@ -0,0 +1,50 @@
+fun rememberActivities(rpc: WearRPCClient): Pair<Array<Activity>, () -> Unit> {
+    var activities: Array<Activity> by remember { mutableStateOf(arrayOf()) }
+    var activitiesQueryCount by remember { mutableIntStateOf(0) }
+    LaunchedEffect(
+        key1 = activitiesQueryCount,
+        block = {
+            async(Dispatchers.Default) {
+                activities = rpc.queryActivities()
+            }
+        },
+    )
```
nit: The "Kotlin-y" way to write functions that have a callback function parameter in the last position is to have the closure come after the function call. e.g.:

```
LaunchedEffect(activitiesQueryCount) {
    async(Dispatchers.Default) {
        activities = rpc.queryActivities()
    }
)
```

kantahrek commented on wear/src/main/java/com/example/util/simpletimetracker/presentation/remember/tags.kt:
```diff
@@ -0,0 +1,51 @@
+fun rememberTags(rpc: WearRPCClient, activityId: Long): Pair<Array<Tag>, () -> Unit> {
+    var tags: Array<Tag> by remember { mutableStateOf(arrayOf()) }
+    var tagsQueryCount by remember { mutableIntStateOf(0) }
+    LaunchedEffect(
+        key1 = tagsQueryCount,
+        block = {
+            async(Dispatchers.Default) {
+                tags = rpc.queryTagsForActivity(activityId)
+            }
+        },
+    )
```
Same nit as in activities.kt

kantahrek reviewed (approved):
LGTM with nits


EDIT: one comment asking for clarification/insights

kantahrek commented on wear/src/main/java/com/example/util/simpletimetracker/presentation/navigation/MainNavigator.kt:
```diff
@@ -0,0 +1,34 @@
+            })
+        }
+        composable("activities/{id}/tags") {
+            TagsScreen(activityId = it.arguments?.getString("id")?.toLong()!!, onSelectTag = {
+                navController.navigate("activities")
+            })
```
Looking at this again... how does this structure lend itself to setting the current activities at tag select time, considering that the setCurrentActivities API needs to know the full list of current activities? It would seem that the TagsScreen would need to introduce its own currentActivities model object so that it can add the selected activity to the list of current ones, but that would introduce its own set of RPC network calls to get that model object until the RPC client is made stateful with a cache.

kantahrek replied:
This extra set of network calls isn't a problem knowing that the plan is to add that caching eventually. More want to check that my impression of how to develop in this presentation layer aligns with yours.

thehale replied:
Good points!

For this first pass, my focus was solely on foundational navigation so I didn't spend too much time here.

From other projects, I know navigation libraries usually either have support for custom payloads OR have a recommended approach for passing more than primitives from one route to another.

According to Google's docs, it looks like the recommendation is exactly what you identified: [add more performance optimizations in the Data Layer.](https://developer.android.com/jetpack/compose/navigation#retrieving-complex-data)

thehale committed:
style: Make `LaunchedEffect` calls more "Kotlin-y"

Apparently, in Kotlin, when the last parameter to a function is a
callback, you can instead specify the callback in a block following
the function call. 🤯

The result looks much cleaner.

Co-authored-by: kantahrek <152828461+kantahrek@users.noreply.github.com>
